src/util/cert/libcrypto/cert.c

```diff
@@ -288,33 +373,84 @@ static errno_t ec_pub_key_to_ssh(TALLOC_CTX *mem_ctx, EVP_PKEY *cert_pub_key,
 #define SSH_RSA_HEADER "ssh-rsa"
 #define SSH_RSA_HEADER_LEN (sizeof(SSH_RSA_HEADER) - 1)
 
+static int sss_rsa_get_key(EVP_PKEY *cert_pub_key,
+                           BIGNUM **_n, BIGNUM **_e)
+{
+    int ret;
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+    *_n = NULL;
+    ret = EVP_PKEY_get_bn_param(cert_pub_key, OSSL_PKEY_PARAM_RSA_N, _n);
+    if (ret != 1) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    *_e = NULL;
+    ret = EVP_PKEY_get_bn_param(cert_pub_key, OSSL_PKEY_PARAM_RSA_E, _e);
+    if (ret != 1) {
+        ret = EINVAL;
+        goto done;
+    }
+
+#else
+
+    const BIGNUM *tmp_n;
+    const BIGNUM *tmp_e:
+
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+    const RSA *rsa_pub_key = NULL;
+    rsa_pub_key = EVP_PKEY_get0_RSA(cert_pub_key);
+    if (rsa_pub_key == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    RSA_get0_key(rsa_pub_key, tmp_n, tmp_e, NULL);
+#else
+    tmp_n = cert_pub_key->pkey.rsa->n;
+    tmp_e = cert_pub_key->pkey.rsa->e;
+#endif /* OPENSSL_VERSION_NUMBER >= 0x10100000L */
+
+    *_n = BN_dup(tmp_n);
+    if (*_n == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    *_e = BN_dup(tmp_e);
+    if (*_e == NULL) {
+        BN_clear_free(*n);
+        ret = ENOMEM;
+        goto done;
+    }
+
+#endif /* OPENSSL_VERSION_NUMBER >= 0x30000000L */
+
+    ret = EOK;
+
+done:
+    return ret;
+}
+
 static errno_t rsa_pub_key_to_ssh(TALLOC_CTX *mem_ctx, EVP_PKEY *cert_pub_key,
                                   uint8_t **key_blob, size_t *key_size)
 {
     int ret;
     size_t c;
     size_t size;
     uint8_t *buf = NULL;
-    const BIGNUM *n;
-    const BIGNUM *e;
+    BIGNUM *n = NULL;
+    BIGNUM *e = NULL;
     int modulus_len;
     unsigned char modulus[OPENSSL_RSA_MAX_MODULUS_BITS/8];
     int exponent_len;
     unsigned char exponent[OPENSSL_RSA_MAX_PUBEXP_BITS/8];
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
-    const RSA *rsa_pub_key = NULL;
-    rsa_pub_key = EVP_PKEY_get0_RSA(cert_pub_key);
-    if (rsa_pub_key == NULL) {
-        ret = ENOMEM;
+    ret = sss_rsa_get_key(cert_pub_key, &n, &e);
+    if (ret != EOK) {
         goto done;
```

Comment:
Is it safe to do
```
    BN_clear_free(n);
    BN_clear_free(e);
```
in the 'done:' section in this case?
I'd just `return;` here.

Reply:
I changed to code so that it will (hopefully) always be safe to call it.